Feat/palette logo

preda-bogdan opened the pull request:
### Summary
The Logo uses a new React component that supports multiple states for light and dark variants.
It defaults and uses the custom logo that was previously set as a default for both states. If palette switcher is not used the behavior should remain the same as before.

While the palette switcher is active you now have the ability to set different variants for light and dark modes.
Since the Palette Switcher uses JS to manually override the browser preferences I could not leverage the `<picture>` tag `media` properties to only switch based on browser preferences so JS is required for the logo switch.

The component uses a listener that detects when the Palette Switcher is added to the Header so that it enables additional features.

The JS from the Logo template is required so that it uses the correct settings from Customizer as well as reacting to the changes from the Palette Switcher using the selective refresh.

### Will affect visual aspect of the product
NO

### Screenshots
Logo when palette switcher is not active:
![image](https://user-images.githubusercontent.com/23024731/126143513-a21bfa0c-0bca-4686-b819-0829514a1fbb.png)

Logo with palette switcher active (2 variants with different and same logo toggle)
![image](https://user-images.githubusercontent.com/23024731/126143763-8a3139de-5c6c-485a-82c5-0b73315cc8e4.png)

![image](https://user-images.githubusercontent.com/23024731/126143907-46f98459-78fa-4362-8080-a7377e434ad8.png)


### Test instructions
1. Test that the Logo behaves the same when Palett Switcher is not active.
2. Add the Palette Switcher to the Header
3. Check that it still behaves the same and that the `Use one logo for both modes` toggle is present on the Logo component.
4. Toggle `Use one logo for both modes` OFF, check that the Logo changes when toggling Light/Dark with the Palette Switcher 

Closes Codeinwp/neve-pro-addon#1067.

irinelenache commented:
@preda-bogdan I tried to test it but with this PR i can't change the logo at all (https://vertis.d.pr/83jxLM). On a fresh instance, even if i have the palette switcher active/inactive, the logo doesn't work. 
Can be checked here:
```
https://apparent-quokka.jurassic.ninja
user: demo
password: rilw1aecdYZX
```
If i import a starter site, still can't change the logo. (https://vertis.d.pr/Ohiy8X)
Please let me know if i didn't test it correctly.

preda-bogdan commented:
@irinelenache Thank you for checking this out. There was an issue on a new instance that somehow was not handled when I tested. Let me know if now everything is working as expected on your end or if you find other issues.

Thank you!

gutoslv commented:
@preda-bogdan the reported issue by Irinel is solved. I've found a bug that is very unlikely to happen in the real world. When you add the palette switcher, then add a new logo, and toggle on customizer the dark mode, and remove the palette switcher, the logo you see on the customizer is the dark mode one, but it doesn't impact the front-end. I've made a video where I demo what happens: https://vertis.d.pr/wSJWfn

preda-bogdan commented:
@gutoslv Thank you, I will take a look and solve this. Good find!

preda-bogdan commented:
@gutoslv I've made some changes and a bit of cleanup. Now the component will reflect the changes inside Customizer when the Palette Switcher is removed, it updates the state to use the same Logo.

Let me know if this is working as expected or if you find other cases where this can component is not working as intended.

Thank you!

irinelenache commented:
@preda-bogdan 

- [ ] I see that on mobile the logo isn't changing when light/dark mode is switched, is this expected? https://vertis.d.pr/AwVnSG

- [ ] If i add 2 Logo & Site identity components into the header, they will appear as the same logo in customizer https://vertis.d.pr/f2pVIt , also, the logo switch works only for the first component

gutoslv commented:
@irinelenache does the first one happens only on the customizer or happens also on frontend?

irinelenache commented:
@gutoslv 

> @irinelenache does the first one happens only on the customizer or happens also on frontend?

It's happening both in customizer and frontend

preda-bogdan commented:
@irinelenache I've made some changes to also support the additional logo from PRO and make it play nicely, the mobile should work as expected now.

Let me know if everything is working correctly now.

cc: @abaicus @selul 
I did some changes so that the components can register inline scripts required for the component to work on the frontend part.
I created a class that handles this and should ensure that we only enqueue a single inline script per handle.
Do you see any issues with this approach that maybe I've overlooked?

Preferably I would like to also have a way to minify/uglify the parsed JS.
Let me know your thoughts on this here.

Thank you!

irinelenache commented:
@preda-bogdan Overall it works fine 🚀 
Found only something small:

- [ ] When i first add the second Logo & Site identity component on mobile (already set with Light/dark logo for desktop), it shows the old logo in customizer https://vertis.d.pr/4RdsZI . After clicking on the palette switch in customizer and refreshing the page, it works fine. (There isn't any problem on frontend)

irinelenache commented:
@preda-bogdan Tested the feature and didn't find any problem 🚀 Will move this to **Ready to merge** 👍

cristian-ungureanu commented on assets/apps/customizer-controls/src/logo-palette/LogoPaletteComponent.js:
```diff
@@ -0,0 +1,257 @@
+					suggestedHeight={height}
+					flexWidth={flexWidth}
+					flexHeight={flexHeight}
+					title={__('Select Logo', 'neve')}
```
Let's replace this with "Change Logo"

cristian-ungureanu commented on assets/apps/customizer-controls/src/logo-palette/LogoPaletteComponent.js:
```diff
@@ -0,0 +1,257 @@
+												resetAttachment('dark');
+											}}
+										/>
+										<p>{__('Dark Mode', 'neve')}</p>
```
Let's change this to "Dark Palette" . '-'. "Change Logo"

cristian-ungureanu commented on assets/apps/customizer-controls/src/logo-palette/LogoPaletteComponent.js:
```diff
@@ -0,0 +1,257 @@
+										className="add-icon"
+										icon={plusCircleFilled}
+									/>
+									<p>{__('Add Light Logo', 'neve')}</p>
```
Let's change this with  "Light Palette" . - . "Upload Logo"

cristian-ungureanu commented on assets/apps/customizer-controls/src/logo-palette/LogoPaletteComponent.js:
```diff
@@ -0,0 +1,257 @@
+											resetAttachment('light');
+										}}
+									/>
+									<p>{__('Light Mode', 'neve')}</p>
```
Let's change this to "Light Palette" .  '-' . "Change Logo"

cristian-ungureanu commented on assets/apps/customizer-controls/src/logo-palette/LogoPaletteComponent.js:
```diff
@@ -0,0 +1,257 @@
+						suggestedHeight={height}
+						flexWidth={flexWidth}
+						flexHeight={flexHeight}
+						title={__('Select Logo Dark', 'neve')}
```
Let's replace this with "Change Logo"

cristian-ungureanu commented on assets/apps/customizer-controls/src/logo-palette/LogoPaletteComponent.js:
```diff
@@ -0,0 +1,257 @@
+											className="add-icon"
+											icon={plusCircleFilled}
+										/>
+										<p>{__('Add Dark Logo', 'neve')}</p>
```
Let's change this with "Dark Palette" . - . "Upload Logo"

cristian-ungureanu commented on header-footer-grid/Core/Components/Logo.php:
```diff
@@ -79,6 +84,162 @@ public function init() {
 		$this->set_property( 'default_selector', '.builder-item--' . $this->get_id() . ' .site-logo' );
 		$this->set_property( 'has_horizontal_alignment', true );
+
+		// add_action( 'wp_enqueue_scripts', [ $this, 'load_scripts' ] );
```
Please remove this if not needed

cristian-ungureanu commented on header-footer-grid/Core/Components/Logo.php:
```diff
@@ -102,6 +264,39 @@ public function add_settings() {
+				'id'                => self::LOGO,
+				'group'             => $this->get_class_const( 'COMPONENT_ID' ),
+				'tab'               => SettingsManager::TAB_GENERAL,
+				// 'transport'         => 'post' . $this->get_class_const( 'COMPONENT_ID' ),
```
Please remove this

cristian-ungureanu commented on header-footer-grid/Core/Components/Logo.php:
```diff
@@ -102,6 +264,39 @@ public function add_settings() {
+				'transport'         => 'refresh',
+				'sanitize_callback' => array( $this, 'sanitize_logo_json' ),
+				'default'           => wp_json_encode( $default ),
+				'label'             => __( 'Logo base', 'neve' ),
```
I think we can remove this string as it's not displayed anywhere.

cristian-ungureanu commented on header-footer-grid/Core/Components/PaletteSwitch.php:
```diff
@@ -117,7 +132,7 @@ function ( $output ) {
 		if ( $this->is_component_active() || is_customize_preview() ) {
 			wp_add_inline_style( 'neve-style', $this->toggle_style() );
-			wp_add_inline_script( 'neve-script', $this->toggle_script() );
+			// wp_add_inline_script( 'neve-script', $this->toggle_script() );
```
Please remove this if not needed

cristian-ungureanu commented on header-footer-grid/templates/components/component-logo.php:
```diff
@@ -60,14 +67,15 @@
-if ( $should_add_skip_lazy ) {
-	$logo_settings['class'] = 'skip-lazy';
-}
+// $should_add_skip_lazy = apply_filters( 'neve_skip_lazy', true );
```
We can't remove the 'neve_skip_lazy' filter. Is there a reason for this?

cristian-ungureanu commented on header-footer-grid/Core/Components/Logo.php:
```diff
@@ -79,6 +84,162 @@ public function init() {
+		});
+	});
+
+	//setCurrentTheme(theme);
```
Please remove if not needed

abaicus commented on header-footer-grid/Core/Components/Logo.php:
```diff
@@ -102,6 +264,39 @@ public function add_settings() {
 			);
 		}
 
```
@preda-bogdan Not a must if you feel that there's too much to change or technical limitations that wouldn't make it worth it. 

What do you think about using multiple settings for the toggle & dark logo instead of using a JSON? This way we make sure there's nothing to migrate and this won't yield weird results.

We do that for the font variants selector. Essentially you'd have multiple values managed by the same UI control:

https://github.com/Codeinwp/neve/blob/master/inc/customizer/options/typography.php#L109-L116
https://github.com/Codeinwp/neve/blob/master/assets/apps/customizer-controls/src/font-family/FontFamilyComponent.js#L104-L107

abaicus reviewed (changes requested):
@preda-bogdan Looks great! 🚀 Here are some issues I found:

- [ ] Dark logo doesn't allow crop skipping;
- [ ] We add a `cursor:pointer;` to the logo select box;
- [ ] Let's make sure we preserve the aspect ratio of the logo. Maybe display the logos one on top of the other with full width (we can either do that or `display:grid; grid-template-columns: 1fr 1fr;`):
![image](https://user-images.githubusercontent.com/15010186/144579102-9d4feb68-44c5-462c-be34-86b9c7d3fcc7.png)

preda-bogdan commented:
@abaicus & @cristian-ungureanu implemented the requested changes let me know if there are other changes required.
Thank you!
@irinelenache can you take a second look and let me know if everything is working as before?

irinelenache commented:
@preda-bogdan Checked again and everything works as expected 👍 
Can be moved to **Ready to merge** after @cristian-ungureanu and @abaicus review it 🚀

abaicus commented:
@preda-bogdan Thanks for the updates! 🚀 I looked a bit into this and found some things that we can do better:

- [ ] Starter content doesn't insert a logo in the customizer anymore. This is due to the fact that the theme mod slug has changed;

- [ ] The aspect ratio still breaks with wide logos. We should make sure that we don't affect the aspect ratio:

![image](https://user-images.githubusercontent.com/15010186/145179389-67da73c3-13ba-472b-83a7-c37b1f8e0ac8.png)

- [ ] Here are the images I was using:

<img width="280" alt="Dark@2x" src="https://user-images.githubusercontent.com/15010186/145179530-43bff1df-c14a-409e-b4f2-93b83b874b8e.png">
<img width="280" alt="Light@2x" src="https://user-images.githubusercontent.com/15010186/145179531-1761d501-a9ca-4365-8291-2d569025c096.png">

- [ ] Let's make the logos display one under another by removing the `grid-template-columns` and adding a `grid-row-gap` to space it a bit. The reason for this is that the strings have changed and are much longer. We might have some issues with certain languages that translate to really long words: 

![image](https://user-images.githubusercontent.com/15010186/145179837-38357957-872d-4a28-9402-e69df1fcf912.png)

My suggestion for the last point is that we could even move the "Dark Palette" / "Light Palette" above the buttons as a title and keep the "Change Logo" on the buttons themselves.  @cristian-ungureanu what do you think about this?

- [ ] Let's use valid markup for the components (both the remove button and the wrapper should be button tags); Right now we're using a div with an onClick directive, which is an antipattern; You can take a look at how this was handled on the color palette selector as it has a similar UI: 

![image](https://user-images.githubusercontent.com/15010186/145180518-bf362762-296d-4684-9c75-74791a00562b.png)

preda-bogdan commented:
@abaicus I've made some changes and addressed the issues you mentioned. One thing that I did different is to just split the text on new lines and keep the current button position since it is the same with the design and it was already agreed upon. Having them on separate lines should be enough there are 21 chars available per line.

Let me know what you think and if there are other changes required.

Thank you!

cristian-ungureanu commented:
@preda-bogdan it seems there are some errors here. Can you check why the tests are failing?

cristian-ungureanu commented:
I'll update this comment as I do progress with the testing:

- [ ] Cropping an image as full width produce this behavior of the control:

<img width="304" alt="Screenshot 2022-01-11 at 15 26 17" src="https://user-images.githubusercontent.com/9929553/148950888-4ef6f0a2-461a-48d1-a00a-4f6bcde3441d.png">

abaicus reviewed (changes requested):
@preda-bogdan Looks good! 🚀 In addition to some suggestions below, I also found the following things:

- There's a weird effect when hovering over the edges of the .logo-select. This probably happens because it has padding and the inner button doesn't actually fill it: 
![image](https://user-images.githubusercontent.com/15010186/148942175-ce653ca4-2e70-44ee-bb4f-7dbfa3553a0e.png)
 
- Removing the dark logo should probably fall back to the value of the light one, otherwise, it will just show a broken image (both inside the customizer and on frontend):  
![image](https://user-images.githubusercontent.com/15010186/148942359-ddf2c475-ad2e-47f3-8001-e08c364bd87b.png)

preda-bogdan commented on assets/apps/customizer-controls/src/logo-palette/scss/_logo-palette.scss:
```diff
@@ -0,0 +1,103 @@
+  }
+
+  p {
+	position: absolute;
```
this is also used for displaying the label over the image on top of the overlay color. changing this to `flex-direction: column` will make them behave as on the same layer

preda-bogdan commented:
> <img alt="Screenshot 2022-01-11 at 15 30 07" width="299" src="https://user-images.githubusercontent.com/9929553/148951900-e2850a74-21a0-4642-862f-aa75bad2b553.png">
> 
>     * [ ]  For larger images, there is no way to skip crop

This is the behavior for the library component when the size is specified and I can not modify that also allowing for full width images without cropping for the logo section does not make much sense to me.

@cristian-ungureanu @abaicus I finished the requested changes let me know if there are more when you have time.

Thank you!

abaicus reviewed (approved):
Looking good! Thanks @preda-bogdan 🚀

cristian-ungureanu reviewed (approved):
@preda-bogdan the code looks good. Please have a look at the tests so we can move this to QA.

irinelenache commented:
@preda-bogdan Tested again and everything seems to work fine 🚀

preda-bogdan commented:
Thank you @irinelenache if everithing is ok and @cristian-ungureanu also approves I think this can be moved to ready to merged.
